// be-async/index.js
(function(){
    var be = {};

    var root = this, previous_be;
    if (root !== null) previous_be = root.be;

    be.noConflict = function(){
        root.be = previous_be;
        return be; 
    };

    be.once = function(fn){
        var called = false;
        return function(){
            if (!called){ 
                fn.apply(this, arguments);
                called = true;
            }
        }
    };

    be.pairs = function(any){
        var r = [];
        for (var k in any){
            if (any.hasOwnProperty(k)){
                r.push([k, any[k]]);
            }
        }
        return r;
    };

    be.flip = function(table){
        var flipped = [], 
            rows = (table[0]||[]).length, 
            cols;
        while (rows--){
            flipped[rows] = [];
            cols = table.length;
            while(cols--){
                flipped[rows][cols] = table[cols][rows];
            }
        } 
        return flipped;
    };

    be.merge = function(arr){
        var mergeed = [];
        for (var n = 0, l = arr.length; n<l; n++) 
            mergeed = mergeed.concat(arr[n]);
        return mergeed;
    };

    be.await = function(ccy, fns, done){
        fns = fns||[];
        var length = fns.length;

        if (ccy === 0||length === 0){
            return done(null);
        }

        var cursor = 0, running = 0, 
            ended = 0, halt = false;

        (function more(next){
            var procs = ccy === -1 ? length
                : (running <= ccy ? ccy - running : 0); 
            while (procs--) (function(i, next){
                be.nextTick(function(){
                    running++;
                    if (!halt && (i < length||length === -1)){
                        fns[i].call(this, function(err){
                            running--;
                            ended++;
                            if (err){
                                halt = true
                                return next(err);
                            }
                            if (ended === length){
                                return next(null);
                            }
                            if (halt) return;
                            more(next);
                        });
                    }
                });
            })(cursor++, next);
        })(done||Function());
    }; 

    be.curry = function(fn, pos, drop, args){
        var inst = this;
        args = args === undefined ? [] : args;
        return function(){
            var _args = [].slice.call(arguments);
            _args.splice.apply(_args, [pos, drop].concat(args));
            return (fn||Function()).apply(inst, _args);
        }
    };

    be.ccurry = function(fn, pos, drop, args){
        var inst = this;
        return function(){
            var _args = [].slice.call(arguments);
            var next = _args.pop();
            return (fn||Function()).apply(inst, 
                _args.concat(be.curry(next, pos, drop, args)));
        }
    };

    be.unbind = function(fn){
        var args = [].slice.call(arguments, 1);
        return function(inst){
            return be.curry(fn, 0, 0, args)
                .apply(inst, [].slice.call(arguments, 1));
        }
    };

    be.pipe = function(){
        var fns = [].slice.call(arguments).reverse();
        return function(val){
            while (fns.length){
                val = fns.pop().call(this, val);
            }
            return val;
        }; 
    };

    be.wrap = function(arr, fn){
        var fns = [], length = arr.length;
        while (length--) fns[length] = be.curry(fn, 0, 0, [arr[length]]);
        return fns;
    };

    be.waterfall = function(fns, done){
        if (!fns) return;
        var args = new Array(0), 
            pairs = be.pairs(fns), 
            length = pairs.length;
        while (length--) pairs[length][1] = (function(pair){
            return function(next){
                fns[pair[0]].apply(this, args.concat(function(err){
                    if (err) return next(err);
                    var _args = [].slice.call(arguments, 1), i = 0;
                    while (i < _args.length) args[i] = _args[i++];
                    var i = _args.length
                    while (_args.length < args.length) args.pop();
                    next(null);
                }));
            }
        })(pairs[length]);
        be.await(1, be.flip(pairs)[1], be.curry(done, 1, 0, args));    
    }; 

    be.parallelLimit = function(fns, limit, done){
        var results = fns.constructor(), 
            pairs = be.pairs(fns), 
            length = pairs.length;
        while (length--) pairs[length][1] = (function(pair){
            return function(next){
                fns[pair[0]].call(this, function(err){
                    if (err) return next(err);
                    var result = [].slice.call(arguments, 1);
                    results[pair[0]] = 
                        result.length === 1 ? result[0] : result;
                    next(null);
                });
            }
        })(pairs[length]);
        be.await(limit, be.flip(pairs)[1], be.curry(done, 1, 0, [results]));    
    };

    be.parallel = be.curry(be.parallelLimit, 1, 0, -1);
    be.series = be.curry(be.parallelLimit, 1, 0, 1);

    be.eachLimit = be.forEachLimit = function(arr, limit, fn, done){ 
        be.await(limit, be.wrap(arr, fn), done);
    };

    be.each = be.forEach = be.curry(be.eachLimit, 1, 0, -1);
    be.eachSeries = be.forEachSeries = be.curry(be.eachLimit, 1, 0, 1);

    be.mapLimit = function(arr, limit, fn, done){
        be.parallelLimit(be.wrap(arr, fn), limit, done);
    };

    be.map = be.curry(be.mapLimit, 1, 0, -1);
    be.mapSeries = be.curry(be.mapLimit, 1, 0, 1);

    be.reduce = be.inject = be.foldl = function(arr, memo, fn, done){
        be.await(1, be.wrap(arr, function(v, next){
            fn(memo, v, function(err, r){ if (!err) memo = r; next(err) })
        }), function(err){ (done||Function())(err, memo) });    
    };

    be.reduceRight = be.foldr = function(arr, memo, fn, done){
        var reversed = [], length = arr.length, i = length;
        while (i--) reversed[length - i - 1] = arr[i];
        be.reduce(reversed, memo, fn, done);
    };

    be.sortBy = function(arr, fn, done){
        be.parallel(be.wrap(arr, fn), function(err, acc){
            (done||Function())(err, be.flip(be.flip([arr, acc]).sort(function(a, b){ 
                return a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0 }))[0])})
    };

    be.filterLimit = be.selectLimit = function(arr, limit, fn, done){
        be.parallelLimit(be.wrap(arr, fn), limit, function(err, acc){
            (done||Function())(err, (function(arr){
                var results = [], length = arr.length, index = 0;
                while (index++ < length) if (arr[index-1][1]) results.push(arr[index-1][0]); 
                return results;
            })(be.flip([arr, acc])));
        });
    };

    be.filter = be.select = be.curry(be.filterLimit, 1, 0, -1); 
    be.filterSeries = be.selectSeries = be.curry(be.filterLimit, 1, 0, 1);
    
    be.rejectLimit = function(arr, limit, fn, done){
        be.filterLimit(arr, limit, function(v, next){
            fn(v, function(err, r){ next(err, !r) })
        }, done);
    }; 

    be.reject = be.select = be.curry(be.rejectLimit, 1, 0, -1); 
    be.rejectSeries = be.selectSeries = be.curry(be.rejectLimit, 1, 0, 1);

    be.detectLimit = function(arr, limit, fn, done){
        done = be.once(done);
        be.await(limit, be.wrap(arr, function(v, next){
            fn(v, function(err, r){ if (r) return done(null, v); next(null) })
        }), be.curry(done, 1, 0, null));
    };

    be.detect = be.select = be.curry(be.detectLimit, 1, 0, -1); 
    be.detectSeries = be.selectSeries = be.curry(be.detectLimit, 1, 0, 1);

    be.concatLimit = function(arr, limit, fn, done){
        be.parallelLimit(be.wrap(arr, fn), limit, function(err, acc){
            (done||Function())(err, be.merge(acc));
        });
    };

    be.concat = be.curry(be.concatLimit, 1, 0, -1); 
    be.concatSeries = be.curry(be.concatLimit, 1, 0, 1);

    be.some = be.any = function(arr, fn, done){
        done = be.once(done);
        be.await(-1, be.wrap(arr, function(v, next){
            fn(v, function(err, r){ if (r) return done(null, !!r); next(null) })
        }), be.curry(done, 1, 0, false));
    };

    be.all = be.every = function(arr, fn, done){
        done = be.once(done);
        be.await(-1, be.wrap(arr, function(v, next){
            fn(v, function(err, r){ if (!r) return done(null, !!r); next(null) })
        }), be.curry(done, 1, 0, true));
    };

    be.nextTick = typeof process !== 'undefined' && process.nextTick ? 
        process.nextTick : typeof setImmediate !== 'undefined' ?  
        setImmediate.bind(null) : function(f){ setTimeout(f, 0) };

    be.Emitter = (function(){
        var events, index;
        var Emitter = function(){
            if (!(this instanceof Emitter)) return new Emitter;
            this.events = {};
        };
        Emitter.prototype = {
            on: function(type, fn){
               return (this.events[type] = this.events[type]||[]).push(fn);
            },
            once: function(type, fn){
                return this.on(type, function _fn(){
                    fn.apply(this, arguments);
                    this.off(type, _fn);
                });
            },
            off: function(type, fn) {
                for (events = this.events[type]||[], index = events.length; index--;) 
                    fn == events[index] && events.splice(index, 1);
                if (!events.length) delete this.events[type];
                return events.length;
            },
            emit: function(type) {
                for (events = this.events[type]||[], index = events.length; index-- > 0;)
                    (events[index]||Function()).apply(this, [].slice.call(arguments, 1));
            }
        }; 
        return Emitter;
    }());

    be.Channel = (function(){
        var Channel = function(slots){
            if (!(this instanceof Channel)) return new Channel;
            be.Emitter.call(this);
            if (slots){
                this.slots = slots;
                this._buffer = [];
            }
        };
        var slot = 0;
        Channel.prototype = Object.create(be.Emitter.prototype); 
        Channel.prototype.constructor = Channel;
        Channel.prototype.buffer = function(data){
            slot = slot % this.slots;
            this._buffer[slot++ % this.slots] = data;
        };
        Channel.prototype.flush = function(){
            while (this._buffer.length) {
                this.push.apply(this, this._buffer.shift());
            }
        };
        Channel.prototype.shift = function(next) {
            var bound = this.on('data', next); 
            if (this.slots) this.flush.call(this);
            return bound;
        };
        var handler = 0;
        Channel.prototype.push = function(){
            var args = [].slice.call(arguments);
            if (!args.length) return !!(this.events = {});
            var handlers = this.events && this.events['data'];
            handler = handler % (handlers||[1]).length;
            if (!handlers && this.slots) return !this.buffer.call(this, args);
            return !!(handlers||[Function()])[handler++].apply(this, args)||true;
        }; 
        return Channel;
    }());

    if (typeof module !== 'undefined' && module.exports) {
        module.exports = be;
    } else if (typeof define !== 'undefined' && define.amd) {
        define([], function(){ return be });
    } else {
        root.be = be;
    } 
})();
